mono/metadata/w32process.c

```diff
@@ -124,132 +137,92 @@ get_process_module_class (void)
 }
 
 static guint32
-unicode_chars (const gunichar2 *str)
+unicode_chars (const gunichar2 *str) // FIXME g_utf16_len
 {
 	guint32 len;
 	for (len = 0; str [len] != '\0'; ++len) {}
 	return len;
 }
 
 static void
-process_set_field_object (MonoObject *obj, const gchar *fieldname, MonoObject *data)
+process_set_field (MonoObjectHandle obj, const char *fieldname, gconstpointer val, gsize size)
 {
-	MonoClass *klass;
-	MonoClassField *field;
-
-	LOGDEBUG (g_message ("%s: Setting field %s to object at %p", __func__, fieldname, data));
-
-	klass = mono_object_class (obj);
+	MonoClass *klass = mono_handle_class (obj);
 	g_assert (klass);
 
-	field = mono_class_get_field_from_name_full (klass, fieldname, NULL);
+	MonoClassField *field = mono_class_get_field_from_name_full (klass, fieldname, NULL);
 	g_assert (field);
 
-	mono_gc_wbarrier_generic_store (((char *)obj) + field->offset, data);
+	gpointer p = field->offset + (char*)MONO_HANDLE_RAW (obj);
+
+	if (size)
+		memmove (p, val, size);
+	else
+		mono_gc_wbarrier_generic_store (p, (MonoObject*)val);
 }
 
 static void
-process_set_field_string (MonoObject *obj, const gchar *fieldname, const gunichar2 *val, guint32 len, MonoError *error)
+process_set_field_object (MonoObjectHandle obj, const char *fieldname, MonoObjectHandle data)
 {
-	MonoDomain *domain;
-	MonoClass *klass;
-	MonoClassField *field;
-	MonoString *string;
-
-	error_init (error);
-
-	LOGDEBUG (g_message ("%s: Setting field %s to [%s]", __func__, fieldname, g_utf16_to_utf8 (val, len, NULL, NULL, NULL)));
+	LOGDEBUG (g_message ("%s: Setting field %s to object at %p", __func__, fieldname, MONO_HANDLE_RAW (data)));
+	process_set_field (obj, fieldname, MONO_HANDLE_RAW (data), 0);
```

Comment:
This raw pointer is a bit long-lived. We might want to pass the handle as another parameter and make it raw over a shorter interval.